driver-core/src/main/com/mongodb/internal/connection/SocketStreamHelper.java

```diff
@@ -78,19 +107,21 @@ static void initialize(final Socket socket, final InetSocketAddress inetSocketAd
         socket.connect(inetSocketAddress, settings.getConnectTimeout(MILLISECONDS));
     }
 
-    @SuppressWarnings("unchecked")
-    private static void setExtendedSocketOptions(final Socket socket) {
-        if (Arrays.stream(ExtendedSocketOptions.class.getDeclaredFields()).anyMatch(f -> f.getName().equals(TCP_KEEPCOUNT))) {
-            try {
-                Method setOptionMethod = Socket.class.getMethod("setOption", SocketOption.class, Object.class);
-                setOptionMethod.invoke(socket, ExtendedSocketOptions.class.getDeclaredField(TCP_KEEPCOUNT).get(null),
-                        TCP_KEEPCOUNT_LIMIT);
-                setOptionMethod.invoke(socket, ExtendedSocketOptions.class.getDeclaredField(TCP_KEEPIDLE).get(null),
-                        TCP_KEEPIDLE_DURATION);
-                setOptionMethod.invoke(socket, ExtendedSocketOptions.class.getDeclaredField(TCP_KEEPINTERVAL).get(null),
-                        TCP_KEEPINTERVAL_DURATION);
-            } catch (Throwable t) {
+    static void setExtendedSocketOptions(final Socket socket) {
+        try {
+            if (SET_OPTION_METHOD != null) {
+                if (KEEP_COUNT_OPTION != null) {
+                    SET_OPTION_METHOD.invoke(socket, KEEP_COUNT_OPTION, TCP_KEEPCOUNT_LIMIT);
+                }
+                if (KEEP_IDLE_OPTION != null) {
+                    SET_OPTION_METHOD.invoke(socket, KEEP_IDLE_OPTION, TCP_KEEPIDLE_DURATION);
+                }
+                if (KEEP_INTERVAL_OPTION != null) {
+                    SET_OPTION_METHOD.invoke(socket, KEEP_INTERVAL_OPTION, TCP_KEEPINTERVAL_DURATION);
+                }
             }
+        } catch (IllegalAccessException | InvocationTargetException e) {
```

Comment:
The method `setExtendedSocketOptions` is "best effort" only with regard to working around access issues and unavailability of a method/class/fields. If everything is present and the code does have access to invoke the `setOption` method, then this method throwing an exception is a problem that must not be ignored, but must be treated like any other exception happening when we set other socket options in a more traditional way.

Reply:
Yeah, I noticed that too.  I was trying not to change the existing behavior, which also ignores this exception, just in case there is a problem calling this method that we're not aware.  I don't want to regress by throwing where we didn't throw before. But maybe that's unlikely.  One option is to log the error but still ignore it. What do you think?